### PR TITLE
Admin Router: Conditional request buffering for /service endpoint

### PR DESCRIPTION
**Pull request by vespian**
## High-level description

This PR makes it possible to configure `/service` endpoint request buffering basing on the `DCOS_SERVICE_REQUEST_BUFFERING` label. See the Jira issue for details.

## Corresponding DC/OS tickets (obligatory)

These DC/OS JIRA ticket(s) must be updated (ideally closed) in the moment this PR lands:

  - https://jira.mesosphere.com/browse/DCOS-20269 `Request entity too large error`

## Related PRs

EE PR: https://github.com/mesosphere/dcos-enterprise/pull/2587

### CODE REVIEW

**Comment by vespian** [comment]
@mesosphere-mergebot bump-ee

**Comment by vespian** [comment]
@mesosphere-mergebot bump-ee

**Comment by vespian** [comment]
@mesosphere-mergebot bump-ee

**Comment by vespian** [comment]
@mesosphere-mergebot bump-ee

**Comment by vespian** [comment]
@mesosphere-mergebot bump-ee

**Comment by vespian** [comment]
@mesosphere-mergebot label Ready for review

**Comment by vespian** [comment]
@sambatyon @jgehrcke @adamtheturtle  Please have a look

File: packages/adminrouter/extra/src/README.md [hunk 1]
```diff
@@ -521,6 +521,48 @@ It is possible to workaround it by using another another HTTP reverse proxy in t
...
 
+### Disabling request buffering for selected applications
+Some applications can not tolerate request buffering while there are others that require it in order to function properly. To support both groups, support for `DCOS_SERVICE_REQUEST_BUFFERING` label was introduced.
+
+If a task has `DCOS_SERVICE_REQUEST_BUFFERING` label set to `false` (string) or `false` (boolean), Admin Router will not perform request buffering for this application. Please see below for an example of how to enable it in an application definition:
+
+```json
+{
+  "labels": {
+    "DCOS_SERVICE_REQUEST_BUFFERING": false
```
Comment (jgehrcke) [inline packages/adminrouter/extra/src/README.md]: GitHub's highlighting shows that this lacks a comma :)

File: packages/adminrouter/extra/src/includes/server/master.conf [hunk 2]
```diff
@@ -172,32 +172,57 @@ location ~ ^/service/(?<serviceid>[0-9a-zA-Z-.]+)$ {
...
         util.clear_dcos_cookies();
-    }
 
-    include includes/proxy-headers.conf;
-    include includes/websockets.conf;
+        -- A hack to enable the control of proxy_request_buffering via Marathon
+        -- labels. The proxy_request_buffering directive by itself is unable to
+        -- parse lua variables directly.
+        if service.request_buffering_required(service_name, marathon_cache) == false then
+            return ngx.exec("@service_requnbuffered");
```
Comment (jgehrcke) [inline packages/adminrouter/extra/src/includes/server/master.conf]: Happy that we got this working!!

Maybe add a comment here saying

```
# Do an internal redirect; proceed handling this request in the 
# @service_requnbuffered location block below.
```

File: packages/adminrouter/extra/src/test-harness/tests/test_service.py [hunk 3]
```diff
@@ -897,3 +897,81 @@ def test_if_requrl_rewriting_can_be_configured(
...
+                            aux_data=new_apps)
+
+        # In theory it is possible to write a test that really checks if the
+        # request was buffered or not. It would require talking to the mocked
+        # endpoint during the test and checking if it is receiving the data as
+        # it is being sent (there is no buffering) or only after the whole
+        # request has been uploaded (Nginx buffers the data). Such a feature
+        # would introduce some extra complexity into the test harness. Simply
+        # checking if AR is printing the warning to the error log seems to be
+        # good enough.
```
Comment (jgehrcke) [inline packages/adminrouter/extra/src/test-harness/tests/test_service.py]: I agree with that. Thanks for the code comment!!

**Comment by vespian** [comment]
@mesosphere-mergebot bump-ee

**Comment by adamtheturtle** [comment]
Please can you find the relevant `teamcity/dcos/test/docker` issue, and make a judgement call on whether this test suite needs to be seen passing for this PR.

**Comment by vespian** [comment]
`teamcity/dcos/test/docker` CI failure seems like CI infra issue unrelated to the PR. I asked #tools-infra team and in the meantime, I have re-run the CI.

**Comment by vespian** [comment]
@orsenthil  Could you please check why @mergebot still thinks that there are <2 reviews? Thanks!

**Comment by vespian** [comment]
@orsenthil  It worked after @sambatyon re-approving the PR. I do not know why it is so, as there were no changes since the last time he approved it :(

**Comment by orsenthil** [comment]
@vespian / @jgehrcke  - I am not sure why these are not counted. If you notice this, add a note to this JIRA -  https://jira.mesosphere.com/browse/QUALITY-1967

**Pull request merged**